File: static/js/action-parser.js
/**
 * ActionParser - Extracts action tags from LLM output text
 *
 * Tag format: [type:name] or [type:name,param:value,param2:value2]
 * Types: mood, gesture, action, eye, head, body
 *
 * Example:
 *   Input:  "[mood:happy] Hello there! [head:nod] How are you? [eye:wink]"
 *   Output: {
 *     cleanText: "Hello there! How are you?",
 *     actions: [
 *       { type: 'mood', name: 'happy', params: {}, position: 0 },
 *       { type: 'head', name: 'nod', params: {}, position: 14 },
 *       { type: 'eye', name: 'wink', params: {}, position: 28 }
 *     ]
 *   }
 */

class ActionParser {
    constructor() {
        // Match [type:name] or [type:name,key:value,key2:value2]
        // Examples: [mood:happy], [gesture:nod,duration:1.2], [action:greet,intensity:0.8]
        this.tagRegex = /\[(\w+):(\w+)(?:,([^\]]+))?\]/g;
    }

    /**
     * Parse text and extract action tags
     * @param {string} text - Raw text with embedded tags
     * @returns {{ cleanText: string, actions: Array<{type: string, name: string, params: object, position: number}> }}
     */
    parse(text) {
        if (!text || typeof text !== 'string') {
            return { cleanText: text || '', actions: [] };
        }

        const actions = [];
        let cleanText = text;
        let match;
        let offset = 0;

        // Reset regex state
        this.tagRegex.lastIndex = 0;

        while ((match = this.tagRegex.exec(text)) !== null) {
            const [fullMatch, type, name, paramsStr] = match;
            const originalPosition = match.index;

            // Parse optional parameters (key:value,key2:value2)
            const params = this._parseParams(paramsStr);

            actions.push({
                type: type.toLowerCase(),
                name: name.toLowerCase(),
                params,
                position: originalPosition - offset  // Position in clean text
            });

            // Track offset for position calculation
            offset += fullMatch.length;
        }

        // Remove all tags from text
        cleanText = text.replace(this.tagRegex, '').trim();
        // Clean up extra spaces
        cleanText = cleanText.replace(/\s+/g, ' ').trim();

        return { cleanText, actions };
    }

    /**
     * Parse comma-separated key:value pairs
     * @param {string} paramsStr - e.g., "intensity:0.8,duration:1.2"
     * @returns {object}
     */
    _parseParams(paramsStr) {
        if (!paramsStr) return {};

        const params = {};
        const pairs = paramsStr.split(',');

        for (const pair of pairs) {
            const [key, value] = pair.split(':').map(s => s.trim());
            if (key && value !== undefined) {
                // Try to parse as number, otherwise keep as string
                const numValue = parseFloat(value);
                params[key.toLowerCase()] = isNaN(numValue) ? value : numValue;
            }
        }

        return params;
    }

    /**
     * Quick check if text contains any action tags
     * @param {string} text
     * @returns {boolean}
     */
    hasActions(text) {
        if (!text) return false;
        this.tagRegex.lastIndex = 0;
        return this.tagRegex.test(text);
    }

    /**
     * Strip all action tags from text (simple version of parse)
     * @param {string} text
     * @returns {string}
     */
    stripTags(text) {
        if (!text) return '';
        return text.replace(this.tagRegex, '').replace(/\s+/g, ' ').trim();
    }

    /**
     * Parse text and estimate timing based on character count before each tag
     * @param {string} text - Raw text with embedded tags
     * @param {number} msPerChar - Milliseconds per character (default 60ms ~150 wpm)
     * @returns {{ cleanText: string, actions: Array<{type: string, name: string, params: object, position: number, delay: number}> }}
     */
    parseWithTiming(text, msPerChar = 60) {
        if (!text || typeof text !== 'string') {
            return { cleanText: text || '', actions: [] };
        }

        const actions = [];
        let cleanText = '';
        let lastIndex = 0;
        let charCount = 0;
        let match;

        // Reset regex state
        this.tagRegex.lastIndex = 0;

        while ((match = this.tagRegex.exec(text)) !== null) {
            const [fullMatch, type, name, paramsStr] = match;

            // Count clean text characters before this tag
            const textBefore = text.slice(lastIndex, match.index);
            const cleanBefore = textBefore.replace(/\s+/g, ' ');
            charCount += cleanBefore.length;
            cleanText += cleanBefore;

            // Parse optional parameters
            const params = this._parseParams(paramsStr);

            // Delay based on characters read so far
            actions.push({
                type: type.toLowerCase(),
                name: name.toLowerCase(),
                params,
                position: match.index,
                delay: Math.floor(charCount * msPerChar)
            });

            lastIndex = match.index + fullMatch.length;
        }

        // Add remaining text
        cleanText += text.slice(lastIndex);
        cleanText = cleanText.replace(/\s+/g, ' ').trim();

        return { cleanText, actions };
    }

    /**
     * Schedule actions to execute at their delay times
     * @param {Array} actions - Actions with delay from parseWithTiming
     * @param {function} executeAction - Callback to execute action: (type, name, params) => void
     * @returns {Array<number>} - Array of timeout IDs (for cancellation)
     */
    scheduleActions(actions, executeAction) {
        const timeouts = [];

        for (const action of actions) {
            const delay = action.delay || 0;
            const timeoutId = setTimeout(() => {
                executeAction(action.type, action.name, action.params);
            }, delay);
            timeouts.push(timeoutId);
        }

        return timeouts;
    }

    /**
     * Cancel scheduled actions
     * @param {Array<number>} timeouts - Timeout IDs from scheduleActions
     */
    cancelScheduled(timeouts) {
        if (timeouts) {
            timeouts.forEach(id => clearTimeout(id));
        }
    }
}

// Export for use in other modules
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { ActionParser };
}

File: vtuber_engine/utils/action_parser.py
"""
Action Parser - Extracts action tags from LLM output text.

Tag format: [type:name] or [type:name,param:value,param2:value2]
Types: mood, gesture, action, eye, head, brow, move, zoom, body

Movement tags:
- [move:up], [move:down], [move:left], [move:right] - Directional movement
- [zoom:in], [zoom:out], [zoom:normal] - Scale avatar
- [body:left], [body:right], [body:front] - Body orientation

Example:
    Input:  "[mood:happy] Hello there! [head:nod] How are you? [eye:wink]"
    Output: {
        "clean_text": "Hello there! How are you?",
        "actions": [
            {"type": "mood", "name": "happy", "params": {}, "position": 0},
            {"type": "head", "name": "nod", "params": {}, "position": 14},
            {"type": "eye", "name": "wink", "params": {}, "position": 28}
        ]
    }

Movement Example:
    Input:  "[move:left] Let me show you this! [zoom:in] Check it out!"
    Output: {
        "clean_text": "Let me show you this! Check it out!",
        "actions": [
            {"type": "move", "name": "left", "params": {}, "position": 0},
            {"type": "zoom", "name": "in", "params": {}, "position": 22}
        ]
    }
"""
import re
from typing import Dict, List, Optional, Any
from dataclasses import dataclass


@dataclass
class ParsedAction:
    """A single parsed action from text."""
    type: str
    name: str
    params: Dict[str, Any]
    position: int
    delay_ms: int = 0  # Delay based on text position


class ActionParser:
    """
    Parser for extracting action tags from LLM output.
    Ported from JavaScript action-parser.js.
    """

    # Match [type:name] or [type:name,key:value,key2:value2]
    TAG_PATTERN = re.compile(r'\[(\w+):(\w+)(?:,([^\]]+))?\]')

    def __init__(self):
        pass

    def parse(self, text: str) -> Dict[str, Any]:
        """
        Parse text and extract action tags.

        Args:
            text: Raw text with embedded tags

        Returns:
            Dict with clean_text and actions list
        """
        if not text or not isinstance(text, str):
            return {"clean_text": text or "", "actions": []}

        actions = []
        offset = 0

        for match in self.TAG_PATTERN.finditer(text):
            full_match = match.group(0)
            action_type = match.group(1).lower()
            name = match.group(2).lower()
            params_str = match.group(3)

            # Parse optional parameters
            params = self._parse_params(params_str)

            actions.append(ParsedAction(
                type=action_type,
                name=name,
                params=params,
                position=match.start() - offset
            ))

            offset += len(full_match)

        # Remove all tags from text
        clean_text = self.TAG_PATTERN.sub('', text).strip()
        # Clean up extra spaces
        clean_text = re.sub(r'\s+', ' ', clean_text).strip()

        return {
            "clean_text": clean_text,
            "actions": [
                {"type": a.type, "name": a.name, "params": a.params, "position": a.position, "delay_ms": a.delay_ms}
                for a in actions
            ]
        }

    def _parse_params(self, params_str: Optional[str]) -> Dict[str, Any]:
        """
        Parse comma-separated key:value pairs.

        Args:
            params_str: e.g., "intensity:0.8,duration:1.2"

        Returns:
            Dict of parameters
        """
        if not params_str:
            return {}

        params = {}
        pairs = params_str.split(',')

        for pair in pairs:
            parts = pair.split(':')
            if len(parts) >= 2:
                key = parts[0].strip().lower()
                value = parts[1].strip()

                # Try to parse as number
                try:
                    params[key] = float(value)
                except ValueError:
                    params[key] = value

        return params

    def has_actions(self, text: str) -> bool:
        """Quick check if text contains any action tags."""
        if not text:
            return False
        return bool(self.TAG_PATTERN.search(text))

    def strip_tags(self, text: str) -> str:
        """Strip all action tags from text."""
        if not text:
            return ""
        result = self.TAG_PATTERN.sub('', text)
        return re.sub(r'\s+', ' ', result).strip()

    # Movement action types
    MOVEMENT_TYPES = {"move", "zoom", "body"}

    # Valid movement values
    MOVEMENT_VALUES = {
        "move": {"up", "down", "left", "right"},
        "zoom": {"in", "out", "normal"},
        "body": {"left", "right", "front"}
    }

    def is_movement_action(self, action_type: str) -> bool:
        """Check if action type is a movement command."""
        return action_type.lower() in self.MOVEMENT_TYPES

    def categorize_actions(self, actions: List[Dict[str, Any]]) -> Dict[str, List[Dict[str, Any]]]:
        """
        Categorize parsed actions into groups.

        Returns:
            Dict with keys: 'animations' (mood/gesture), 'movements' (move/zoom/body)
        """
        result = {
            "animations": [],
            "movements": []
        }

        for action in actions:
            action_type = action.get("type", "").lower()
            if self.is_movement_action(action_type):
                result["movements"].append(action)
            else:
                result["animations"].append(action)

        return result

    def parse_with_timing(self, text: str, ms_per_char: int = 60) -> Dict[str, Any]:
        """
        Parse text and estimate timing based on character count before each tag.

        Args:
            text: Raw text with embedded tags
            ms_per_char: Milliseconds per character (~60ms = 150 wpm)

        Returns:
            Dict with clean_text and actions with delay_ms
        """
        if not text or not isinstance(text, str):
            return {"clean_text": text or "", "actions": []}

        actions = []
        clean_text = ""
        last_index = 0
        char_count = 0

        for match in self.TAG_PATTERN.finditer(text):
            # Count clean text characters before this tag
            text_before = text[last_index:match.start()]
            clean_before = re.sub(r'\s+', ' ', text_before)
            char_count += len(clean_before)
            clean_text += clean_before

            full_match = match.group(0)
            action_type = match.group(1).lower()
            name = match.group(2).lower()
            params_str = match.group(3)

            params = self._parse_params(params_str)

            actions.append(ParsedAction(
                type=action_type,
                name=name,
                params=params,
                position=match.start(),
                delay_ms=int(char_count * ms_per_char)
            ))

            last_index = match.end()

        # Add remaining text
        clean_text += text[last_index:]
        clean_text = re.sub(r'\s+', ' ', clean_text).strip()

        return {
            "clean_text": clean_text,
            "actions": [
                {"type": a.type, "name": a.name, "params": a.params, "position": a.position, "delay_ms": a.delay_ms}
                for a in actions
            ]
        }
